server/src/main/java/org/elasticsearch/cluster/coordination/Coordinator.java

```diff
@@ -400,30 +401,37 @@ private void handleJoinRequest(JoinRequest joinRequest, JoinHelper.JoinCallback
                 JoinTaskExecutor.ensureMajorVersionBarrier(joinRequest.getSourceNode().getVersion(),
                     stateForJoinValidation.getNodes().getMinNodeVersion());
             }
+            sendValidateJoinRequest(stateForJoinValidation, joinRequest, joinCallback);
 
-            // validate the join on the joining node, will throw a failure if it fails the validation
-            joinHelper.sendValidateJoinRequest(joinRequest.getSourceNode(), stateForJoinValidation, new ActionListener<Empty>() {
-                @Override
-                public void onResponse(Empty empty) {
-                    try {
-                        processJoinRequest(joinRequest, joinCallback);
-                    } catch (Exception e) {
-                        joinCallback.onFailure(e);
-                    }
-                }
-
-                @Override
-                public void onFailure(Exception e) {
-                    logger.warn(() -> new ParameterizedMessage("failed to validate incoming join request from node [{}]",
-                        joinRequest.getSourceNode()), e);
-                    joinCallback.onFailure(new IllegalStateException("failure when sending a validation request to node", e));
-                }
-            });
         } else {
             processJoinRequest(joinRequest, joinCallback);
         }
     }
 
+    // public for tests
+    public void sendValidateJoinRequest(ClusterState stateForJoinValidation, JoinRequest joinRequest,
```

Comment:
can you move this test to the same package as coordinator, and make this method package-visible instead?